### ISSUE TITLE
Question about NCCL_ALGO_NVLS algorithm and NVLink SHARP technology

### ISSUE THREAD

**Issue by haswelliris** [NVIDIA/nccl#807]
Hello NCCL team,
I noticed that a new algorithm called NCCL_ALGO_NVLS has been implemented in version 2.17. However, there is no information available about NVLS, even in the NCCL documentation which only lists the Tree, Ring and Collnet algorithms.
After watching the GTC conference yesterday, I believe that NVLink SHARP is the new technology being referred to.
Could you please tell us more information about the NVLink SHARP and the NVLS algorithm?
How does it work and how does NCCL implement it?

**Comment by AddyLaddy** [comment]
Yes, the new NCCL_ALGO=NVLS supports intra-node NVLink SHARP which is a new feature of the NVLink4 NVSwitches.
It is a HW implemented feature that adds the offload of arithmetic to the NVSwitches plus a multicast mechanism. Combining these two allows NCCL to use the NVSwitches to perform an AllReduce in NCCL 2.17.x

https://developer.nvidia.com/blog/upgrading-multi-gpu-interconnectivity-with-the-third-generation-nvidia-nvswitch/

It is available on systems that have H100 and NVLink4 switches.

**Comment by AddyLaddy** [comment]
The new CUDA 12.1 APIs that are used to create and manage Multicast groups are documented here:

https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MULTICAST.html

There are dedicated NCCL kernels for the NVLS feature that make use of load and store instructions to facilitate the HW offloaded operations.

https://github.com/NVIDIA/nccl/blob/5d3ab08b69754cb863ab1bf6e3fb6f7612bb725b/src/collectives/device/all_reduce.h#L375

**Comment by sjeaugey** [comment]
To be clear, you need NCCL compiled against CUDA 12.1 (and a CUDA 12.1-compatible driver) to have NCCL use it, on compatible systems like DGX H100. Allreduce BusBw should jump from ~370 to ~480 GB/s.

The NVLS algorithm is new in 2.17; it is still very much being worked on in the context of multi-node (combining NVLink SHARP with IB SHARP and maybe other algorithms) so the code and documentation should settle in the next releases.

**Comment by dominicshanshan** [comment]
> To be clear, you need NCCL compiled against CUDA 12.1 (and a CUDA 12.1-compatible driver) to have NCCL use it, on compatible systems like DGX H100. Allreduce BusBw should jump from ~370 to ~480 GB/s.
> 
> The NVLS algorithm is new in 2.17; it is still very much being worked on in the context of multi-node (combining NVLink SHARP with IB SHARP and maybe other algorithms) so the code and documentation should settle in the next releases.

hi Sylvain, customer want to test this new feature NVLS algorithm in HGX H800, do you have any suggestion or guideline to enable it?

**Comment by Zha0q1** [comment]
Hey @AddyLaddy do you have a link to the APIs used to offload arithmetic to nvswitches? Very interested in the perf gain with all_gather and reduce_scatter

**Comment by Ryan201802** [comment]
> To be clear, you need NCCL compiled against CUDA 12.1 (and a CUDA 12.1-compatible driver) to have NCCL use it, on compatible systems like DGX H100. Allreduce BusBw should jump from ~370 to ~480 GB/s.
> 
> The NVLS algorithm is new in 2.17; it is still very much being worked on in the context of multi-node (combining NVLink SHARP with IB SHARP and maybe other algorithms) so the code and documentation should settle in the next releases.

Hey @sjeaugey how to understand "in the context of multi-node (combining NVLink SHARP with IB SHARP)"? Which APIs can be called in this situation?

**Comment by stephenmsachs** [comment]
### Issue Cleanup: Helping Us Focus on Current Challenges

  We're reviewing older issues to ensure we prioritize the most relevant and active ones. Since this issue hasn't seen updates in over 6 months, we'll be closing it for now.

  ,*This change helps us focus our efforts on addressing any current issues our users are facing.* If this issue still affects you, please don't hesitate to reopen it with a quick update (e.g., "Still relevant on [version=X]").
  Thanks for your understanding and for contributing to NCCL.
